### ISSUE TITLE
Hook to intercept engine chat messages

### ISSUE THREAD

**Issue by Earu** [Facepunch/garrysmod-requests#1857]
So I ran into this issue a while ago, I have my own networking system for a chat addon I've been working on and whenever someone uses the source chat or the console commands its completely by-passing my networking which is a problem, I can't really use PlayerSay to achieve that because it needs to be handled pre-PlayerSay.

To be precise the issues I'm facing are that I'm handling the gamemode calls myself (PlayerSay, etc...) and so I need to catch the result of said hooks to follow up on my networking code, I cannot do that with the hooks currently available not to mention if I was using PlayerSay, other hooks might be executed before mine etc...

TL;DR not much control over the default chat networking,
So I think it could be quite beneficial to have a hook that does this.

**Comment by robotboy655** [comment]
So, you asking for a serverside hook that is run just before PlayerSay so you can have your chat addon? ...or what?

> not to mention if I was using PlayerSay, other hooks might be executed before mine etc...

No matter how many hooks are added you will run into this problem.

**Comment by Earu** [comment]
A hook, or whatever way to handle the networking of messages yourself really, I'm just stating my case it could be used for other things well.

It could just be like a global callback that you could override for instance.
